### PR TITLE
Ensure `ApplyData.EvalDelta` is clear if `applyEvalDelta` fails

### PR DESCRIPTION
**Pull request by algomaxj**
If `StatefulEval` returns an invalid `EvalDelta` for a `ClearStateProgram`, the current code would still fill in `ApplyData` with the `EvalDelta`, which it shouldn't do.

Fortunately, this case should be unreachable, because `StatefulEval` should never return an invalid `EvalDelta`. However, I used this as an opportunity to clean up the relatively confusing `errIfNotApplied` path in favor of a new error type.

This also uncovered a test that wasn't working properly -- it was assuming the `ClearStateProgram` was succeeding because of the non-empty `EvalDelta` in `ApplyData`, when in fact it was failing.

### CODE REVIEW

File: data/transactions/application_test.go [hunk 1]
```diff
@@ -519,22 +512,21 @@ func TestAppCallApplyGlobalStateDeltas(t *testing.T) {
 	// now check errors with non-default values
 	b.SetProto(protocol.ConsensusV23)
 	ed.GlobalDelta["uint"] = basics.ValueDelta{Action: basics.SetUintAction, Uint: 1}
-	err = ac.applyEvalDelta(ed, params, creator, sender, &b, appIdx, errIfNotApplied)
-	a.NoError(err)
+	err = ac.applyEvalDelta(ed, params, creator, sender, &b, appIdx)
+	a.True(isApplyError(err))
```
Comment (algorandskiy) [inline data/transactions/application_test.go:516]: nit: add `a.Error(err)` above

File: data/transactions/application_test.go [hunk 2]
```diff
@@ -583,18 +575,19 @@ func TestAppCallApplyLocalsStateDeltas(t *testing.T) {
 	var params basics.AppParams
 	var appIdx basics.AppIndex
 	var b testBalances
-	var errIfNotApplied = false
+	errIfNotApplied := false
```
Comment (algorandskiy) [inline data/transactions/application_test.go:578]: is it needed?

File: data/transactions/application_test.go [hunk 3]
```diff
@@ -1012,6 +1004,7 @@ func TestAppCallClearState(t *testing.T) {
 	ad := &ApplyData{}
 	b.appCreators = make(map[basics.AppIndex]basics.Address)
 	b.balances = make(map[basics.Address]basics.AccountData, 2)
+	b.SetProto(protocol.ConsensusFuture)
```
Comment (algorandskiy) [inline data/transactions/application_test.go:1007]: good catch

File: data/transactions/application_test.go (hunk 1 above)
Comment (algomaxj) [reply data/transactions/application_test.go:519]: done (here and below as well)

File: data/transactions/application_test.go (hunk 2 above)
Comment (algomaxj) [reply data/transactions/application_test.go]: nope! removed

**Pull request merged**